subprojects/logging/src/main/java/org/gradle/internal/logging/sink/ProgressLogEventGenerator.java

```diff
@@ -45,169 +50,174 @@
-                    }
-                    break;
-                case HeaderStarted:
-                    listener.onOutput(plainTextEvent(startTime, EOL));
-                    break;
-                case HeaderCompleted:
-                    return;
-                default:
-                    throw new IllegalStateException("state is " + state);
+        void flushOutput(long completeTime) {
+            if (!bufferedLogs.isEmpty()) {
+                if (lastRenderedBuildOpId != buildOpIdentifier) {
+                    listener.onOutput(newLine(startTime));
+                    listener.onOutput(header(loggingHeader));
+                    lastRenderedBuildOpId = buildOpIdentifier;
+                }
+
+                for (RenderableOutputEvent renderableEvent : bufferedLogs) {
+                    listener.onOutput(renderableEvent);
+                }
             }
-            state = State.HeaderCompleted;
-        }
 
-        public void complete() {
-            boolean hasStatus = GUtil.isTrue(status);
-            switch (state) {
-                case None:
-                    if (hasLoggingHeader && hasStatus) {
-                        doOutput(styledTextEvent(completeTime,
-                            new StyledTextOutputEvent.Span(loggingHeader + ' '),
-                            new StyledTextOutputEvent.Span(Style.ProgressStatus, status),
-                            new StyledTextOutputEvent.Span(EOL)));
-                    } else if (hasLoggingHeader) {
-                        doOutput(plainTextEvent(completeTime, loggingHeader + EOL));
-                    }
-                    break;
-                case HeaderStarted:
-                    assert hasLoggingHeader;
-                    if (hasStatus) {
-                        doOutput(styledTextEvent(completeTime,
-                            new StyledTextOutputEvent.Span(" "),
-                            new StyledTextOutputEvent.Span(Style.ProgressStatus, status),
-                            new StyledTextOutputEvent.Span(EOL)));
-                    } else {
-                        doOutput(plainTextEvent(completeTime, EOL));
-                    }
-                    break;
-                case HeaderCompleted:
-                    if (hasLoggingHeader && hasStatus) {
-                        doOutput(styledTextEvent(completeTime,
-                            new StyledTextOutputEvent.Span(loggingHeader + ' '),
-                            new StyledTextOutputEvent.Span(Style.ProgressStatus, status),
-                            new StyledTextOutputEvent.Span(EOL)));
-                    }
-                    break;
-                default:
-                    throw new IllegalStateException("state is " + state);
+            // Output task outcome in non-interactive mode
+            if (!interactive && GUtil.isTrue(loggingHeader) && GUtil.isTrue(status)) {
```

Comment:
The comment here is indicative of the fact that the expression should be broken out as method.